### PR TITLE
Add ItemStack key-value meta storage and per-stack descriptions

### PR DESCRIPTION
**Pull request by rubenwardy**
Adds ItemStackMetadata, sharing code with NodeMetadata.
Backwards compatible.

- [x] Fix issues with MetaDataRef::checkobject()
- [x] Actually run and test
- [x] Fix polymorphic destructor warning
- [x] Add docs
- [x] Make getString const ref
- [x] Perstack description
- [x] Fix duplicated l_from_table and l_to_table
- [x] Add ItemStackMetaData to move item serialize from Metadata?
- [ ] Test from_table

### CODE REVIEW

File: src/metadata.cpp [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+{
+	if (var.empty()) {
+		m_stringvars.erase(name);
+	} else {
+		m_stringvars[name] = var;
+	}
+}
+
+std::string Metadata::resolveString(const std::string &str,
+	unsigned short recursion) const
```
Comment (nerzhul) [inline src/metadata.cpp]: u16

File: src/metadata.h [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+	void deSerialize(const std::string &in);
+
+	bool operator==(const Metadata &other) const;
+	inline bool operator!=(const Metadata &other) const
+	{
+		return !(*this == other);
+	}
+
+	// Generic key/value store
+	std::string getString(const std::string &name, unsigned short recursion = 0) const;
```
Comment (nerzhul) [inline src/metadata.h]: maybe return a const ref ?

**Comment by nerzhul** [comment]
please don't copy all data when retriving (getter), please use a std::string *res variable instead to prevent many costly copy operations

**Comment by rubenwardy** [comment]
That code already existed, I just moved it

**Comment by nerzhul** [comment]
yes, i know, but please optimize it, this will benefic to all metas

File: src/inventory.h [hunk 3]
```diff
@@ -32,10 +33,14 @@ struct ToolCapabilities;
 
 struct ItemStack
 {
-	ItemStack(): name(""), count(0), wear(0), metadata("") {}
+	ItemStack(): name(""), count(0), wear(0) {}
 	ItemStack(std::string name_, u16 count_,
-			u16 wear, std::string metadata_,
+			u16 wear, Metadata metadata_,
 			IItemDefManager *itemdef);
+	ItemStack(std::string name_, u16 count_,
```
Comment (nerzhul) [inline src/inventory.h]: const std::string &_name

File: src/metadata.cpp [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+*/
+
+#define DESERIALIZE_START '\x01'
+#define DESERIALIZE_KV_DELIM '\x02'
+#define DESERIALIZE_PAIR_DELIM '\x03'
+#define DESERIALIZE_START_STR "\x01"
+#define DESERIALIZE_KV_DELIM_STR "\x02"
+#define DESERIALIZE_PAIR_DELIM_STR "\x03"
+
+std::string Metadata::serialize() const
```
Comment (nerzhul) [inline src/metadata.cpp]: serialize(std::string *res) (please change all proto which copy std::string as return

File: src/metadata.h [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+	void setString(const std::string &name, const std::string &var);
+	// Support variable names in values
+	std::string resolveString(const std::string &str, unsigned short recursion = 0) const;
+	StringMap getStrings() const
+	{
+		return m_stringvars;
+	}
+protected:
+	StringMap m_stringvars;
```
Comment (nerzhul) [inline src/metadata.h]: StringMap is an ordered map, please use UNORDERED_MAP<std::string, std::string> instead. ordering is not needed there

File: src/metadata.h [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+	{
+		return !(*this == other);
+	}
+
+	// Generic key/value store
+	std::string getString(const std::string &name, unsigned short recursion = 0) const;
+	void setString(const std::string &name, const std::string &var);
+	// Support variable names in values
+	std::string resolveString(const std::string &str, unsigned short recursion = 0) const;
+	StringMap getStrings() const
```
Comment (nerzhul) [inline src/metadata.h]: i think this should be returned as a const ref if not modified by caller, else use a pointed result value

File: src/metadata.h (hunk 5 above)
Comment (rubenwardy) [inline src/metadata.h]: src/util/string.h:58:

```cpp
typedef UNORDERED_MAP<std::string, std::string> StringMap;
```

Comment (nerzhul) [inline src/metadata.h]: right

File: src/metadata.h [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+	void deSerialize(const std::string &in);
+
+	bool operator==(const Metadata &other) const;
+	inline bool operator!=(const Metadata &other) const
+	{
+		return !(*this == other);
+	}
+
+	// Generic key/value store
+	std::string getString(const std::string &name, u16 recursion = 0) const;
```
Comment (nerzhul) [inline src/metadata.h]: i looked at the usage cases it should return a const std::string &

File: src/metadata.h [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+	void setString(const std::string &name, const std::string &var);
+	// Support variable names in values
+	std::string resolveString(const std::string &str, u16 recursion = 0) const;
```
Comment (nerzhul) [inline src/metadata.h]: same here

File: src/guiFormSpecMenu.cpp [hunk 9]
```diff
@@ -2303,7 +2303,14 @@ void GUIFormSpecMenu::drawList(const ListDrawSpec &s, int phase,
 			// Draw tooltip
 			std::wstring tooltip_text = L"";
 			if (hovering && !m_selected_item) {
-				tooltip_text = utf8_to_wide(item.getDefinition(m_client->idef()).description);
+				if (hovering && !m_selected_item) {
+					Metadata *meta = &item.metadata;
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: useless variable

File: src/inventory.cpp [hunk 10]
```diff
@@ -289,7 +290,8 @@ void ItemStack::deSerialize(std::istream &is, IItemDefManager *itemdef)
 				wear = stoi(wear_str);
 
 			// Read metadata
-			metadata = deSerializeJsonStringIfNeeded(is);
+			std::string serializedMeta = deSerializeJsonStringIfNeeded(is);
```
Comment (nerzhul) [inline src/inventory.cpp]: call this directly in deSerialize to prevent a copy

Comment (rubenwardy) [inline src/inventory.cpp]: done

File: src/guiFormSpecMenu.cpp [hunk 11]
```diff
@@ -2303,7 +2303,12 @@ void GUIFormSpecMenu::drawList(const ListDrawSpec &s, int phase,
 			// Draw tooltip
 			std::wstring tooltip_text = L"";
 			if (hovering && !m_selected_item) {
-				tooltip_text = utf8_to_wide(item.getDefinition(m_client->idef()).description);
+				const std::string &desc = item.metadata.getString("description");
+				if (desc.empty())
+					tooltip_text =
```
Comment (nerzhul) [inline src/guiFormSpecMenu.cpp]: it's a good place to use braces to prsent problems

File: src/itemstackmetadata.cpp [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+			it != m_stringvars.end(); ++it) {
+		os2 << it->first << DESERIALIZE_KV_DELIM
+		   << it->second << DESERIALIZE_PAIR_DELIM;
+	}
+	os << serializeJsonStringIfNeeded(os2.str());
+}
+
+void ItemStackMetadata::deSerialize(std::istream &is)
+{
+	std::string in = deSerializeJsonStringIfNeeded(is);
```
Comment (nerzhul) [inline src/itemstackmetadata.cpp]: string copy, please use a reference or a pointer as parameter of deSerialize function for your result

File: src/script/lua_api/l_itemstackmeta.cpp [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+	lua_pop(L, 1);  // drop methodtable
+
+	// Cannot be created from Lua
+	//lua_register(L, className, create_object);
+}
+
+const char ItemStackMetaRef::className[] = "ItemStackMetaRef";
+const luaL_reg ItemStackMetaRef::methods[] = {
+	luamethod(MetaDataRef, get_string),
+	luamethod(MetaDataRef, set_string),
```
Comment (nerzhul) [inline src/script/lua_api/l_itemstackmeta.cpp]: booleans ?

File: src/util/serialize.cpp [hunk 14]
```diff
@@ -354,6 +354,55 @@ std::string deSerializeJsonString(std::istream &is)
 	return os.str();
 }
 
+std::string serializeJsonStringIfNeeded(const std::string &s)
```
Comment (nerzhul) [inline src/util/serialize.cpp]: please stopping returning string, use std::string * as a parameter for your data write

File: src/script/lua_api/l_itemstackmeta.cpp (hunk 13 above)
Comment (rubenwardy) [inline src/script/lua_api/l_itemstackmeta.cpp]: out of scope for this PR

File: src/guiFormSpecMenu.cpp (hunk 11 above)
Comment (rubenwardy) [inline src/guiFormSpecMenu.cpp]: I much prefer to always use braces for blocks - however it's pretty much guaranteed that another dev will ask me to remove them

File: src/util/serialize.cpp (hunk 14 above)
Comment (rubenwardy) [inline src/util/serialize.cpp]: out of scope. This is a util function that wraps around other functions, many other places used this and it should be done separately

Comment (rubenwardy) [inline src/util/serialize.cpp]: I'll do this however, just because it should probably be done

Comment (rubenwardy) [inline src/util/serialize.cpp]: Nevermind, I feel like this is a out of scope non-issue. If mods ever store enough meta to make this slow, they shouldn't be storing that much meta in an itemstack - given that it's stored in the player file

**Comment by sofar** [comment]
Tested and works:



Code used in `minetest_game` to test:

```
diff --git a/mods/default/craftitems.lua b/mods/default/craftitems.lua
index 0c51c713..b123cd4d 100644
--- a/mods/default/craftitems.lua
+++ b/mods/default/craftitems.lua
@@ -94,6 +94,8 @@ minetest.register_on_player_receive_fields(function(player, formname, fields)
                local data_str = minetest.serialize(data)
 
                if new_stack then
+                       local meta = new_stack:get_meta()
+                       meta:set_string("description", fields.title)
                        new_stack:set_metadata(data_str)
                        if inv:room_for_item("main", new_stack) then
                                inv:add_item("main", new_stack)
@@ -101,6 +103,8 @@ minetest.register_on_player_receive_fields(function(player, formname, fields)
                                minetest.add_item(player:getpos(), new_stack)
                        end
                else
+                       local meta = stack:get_meta()
+                       meta:set_string("description", fields.title)
                        stack:set_metadata(data_str)
                end
 
```

(of course, the book should convert the metadata to convert to the new storage instead)

**Comment by sofar** [comment]
This has my full support. :+1:

File: src/guiFormSpecMenu.cpp (hunk 11 above)
Comment (ShadowNinja) [inline src/guiFormSpecMenu.cpp]: No, in this case I think it's quite clear that braces are warranted since there's an else clause.

File: src/itemstackmetadata.cpp [hunk 15]
```diff
@@ -0,0 +1,43 @@
+#include "itemstackmetadata.h"
+#include "util/serialize.h"
+#include "util/strfnd.h"
+
+#define DESERIALIZE_START '\x01'
+#define DESERIALIZE_KV_DELIM '\x02'
+#define DESERIALIZE_PAIR_DELIM '\x03'
+#define DESERIALIZE_START_STR "\x01"
+#define DESERIALIZE_KV_DELIM_STR "\x02"
+#define DESERIALIZE_PAIR_DELIM_STR "\x03"
```
Comment (ShadowNinja) [inline src/itemstackmetadata.cpp]: This serialization method will cause issues with special characters and is hard to edit.  Why not use JSON?

Comment (rubenwardy) [inline src/itemstackmetadata.cpp]: oh crap, I forgot about this

Comment (nerzhul) [inline src/itemstackmetadata.cpp]: i agree with shadowninja that json should be used...
